File: frontend/src/api/executiveDashboard.ts
/**
 * api/executiveDashboard.ts — Executive Dashboard API.
 * Pack 1-5 (full coverage).
 */
import { api as apiClient } from "./client";

// ─── Pack 1 ───
export interface ExecCompanyInSector {
  company_id: string;
  name: string;
  pct: number;
  board_id?: string | null;
  task_total: number;
  task_done: number;
}

export interface ExecSectorRow {
  id: string;
  label: string;
  color: string;
  companies_total: number;
  companies_active: number;
  avg_pct: number;
  companies: ExecCompanyInSector[];
}

export interface ExecBottomMetrics {
  proj_count: number;
  task_count: number;
  done_proj: number;
  done_tasks: number;
  deferred_proj: number;
  deferred_tasks: number;
  avg_completion: number;
}

export interface ExecAvailableSector {
  id: string;
  label: string;
  color: string;
}

// ─── Pack 2 — Ratings ───
export interface ExecRingCard {
  label: string;
  rated_count: number;
  total: number;
  not_covered: number;
  accent: string;
  score: number | null;
  delta_2024: number;
}

export interface ExecRatingCell {
  rating: string | null;
  outlook: string | null;
  score: string | null;
  rated_at: string | null;
  report_url: string | null;
}

export interface ExecRatingRow {
  company_id: string;
  name: string;
  fitch: ExecRatingCell | null;
  sp: ExecRatingCell | null;
  moodys: ExecRatingCell | null;
  sf: ExecRatingCell | null;
  sp_esg: ExecRatingCell | null;
  cdp: ExecRatingCell | null;
}

export interface ExecRatingsBlock {
  ring_cards: ExecRingCard[];
  rows: ExecRatingRow[];
  rated_total_unique: number;
  overall_total: number;
}

// ─── Pack 2 — Execution chart ───
export interface ExecExecutionRow {
  company_id: string;
  name: string;
  pct: number;          // факт: % завершённых задач
  plan_pct?: number;    // план: % задач, чей дедлайн уже наступил
  sector: string;
}

// ═══ Pack 4 — Row 3 ═══

export interface ExecDirectionRow {
  id: string;
  label: string;
  color: string;
  projects_total: number;
  projects_done: number;
  tasks_total: number;
  tasks_done: number;
  progress_pct: number;
}

export interface ExecGovernanceCompany {
  company_id: string;
  name: string;
  sector: string;
  score: number;
  score_pct: number;
  board_size: number;
  independent_count: number;
  women_count: number;
  indep_pct: number;
  women_pct: number;
}

export interface ExecGovernanceBlock {
  total_companies: number;
  avg_score: number;
  top_score: number;
  avg_indep_pct: number;
  avg_women_pct: number;
  top_companies: ExecGovernanceCompany[];
}

export interface ExecStandardsRing {
  done: number;
  active: number;
  init: number;
  not_started: number;
  pct: number;
}

export interface ExecStandardsAttention {
  company_id: string;
  name: string;
  sector: string;
  ifrs_status: string;
  forensic_status: string;
  gaps: string[];
}

export interface ExecStandardsBlock {
  total_companies: number;
  ifrs: ExecStandardsRing;
  forensic: ExecStandardsRing;
  attention_list: ExecStandardsAttention[];
}

// ═══ Pack 5 — Row 2.55 / 2.6 / 2.7 ═══

// Block 1: Economic Effect
export interface ExecEEKpi {
  realized_sum: number;
  planned_sum: number;
  pipeline_sum: number;
  conversion_pct: number;
  done_count: number;
  active_count: number;
  total_count: number;
  has_data: boolean;
}

export interface ExecEEProject {
  project_id: string;
  title: string;
  company_name: string;
  sector: string;
  direction: string | null;
  status: string;
  planned_value: number;
  realized_value: number;
  pct_realized: number;
  unit: string;
}

export interface ExecEconomicEffectBlock {
  year: number;
  kpi: ExecEEKpi;
  top_projects: ExecEEProject[];
}

// Block 2: BP Tracker
export interface ExecBPCompanyRow {
  company_id: string;
  name: string;
  sector: string;
  plan_value: number;
  fact_value: number;
  pct: number | null;
  display_pct: number | null;
  display_label: string | null;
  display_label_full: string | null;
  delta: number | null;
  cls: string;
  note: string | null;
}

export interface ExecBPBlock {
  year: number;
  prev_year: number;
  metric: string;
  metric_label: string;
  standard: string;
  mode: string;
  head_sub: string;
  is_signed_metric: boolean;
  plan_total: number;
  fact_total: number;
  sum_plan_ll: number;
  sum_fact_plan_ll: number;
  sum_prev_ll: number;
  sum_fact_ll: number;
  overall_pct: number | null;
  prev_overall_pct: number | null;
  overall_delta: number | null;
  overall_label: string | null;
  rows: ExecBPCompanyRow[];
  on_target: number;
  attention: number;
  behind: number;
  total_count: number;
  with_pct_count: number;
  available_metrics: string[];
}

// Block 3: Tax Contribution
export interface ExecTaxKpi {
  income_tax: number;
  vat: number;
  total: number;
  yoy_total_pct: number | null;
  yoy_income_tax_pct: number | null;
  yoy_vat_pct: number | null;
  budget_share_pct: number | null;
  budget: number | null;
  vat_is_estimate?: boolean;    // Pack 7.9h: НДС = revenue × 12% оценка
}

export interface ExecTaxTopPayer {
  company_id: string;
  name: string;
  sector: string;
  amount: number;
  share_pct: number;
}

export interface ExecTaxBlock {
  year: number;
  prev_year: number;
  has_data: boolean;
  standard: string;
  cos_count: number;
  missing_companies?: string[]; // Pack 7.9h: компании без NSBU PL
  kpi: ExecTaxKpi;
  top_payers: ExecTaxTopPayer[];
}

// ─── Top-level payload ───
export interface ExecutiveDashboardData {
  year: number;
  total_companies: number;
  title_main: string;
  title_sub: string;
  row1_title: string;
  row1_subtitle: string;
  sectors: ExecSectorRow[];
  bottom_metrics: ExecBottomMetrics;

  // Pack 2
  ratings: ExecRatingsBlock | null;
  execution_chart: ExecExecutionRow[];
  avg_execution_pct: number;

  // Pack 4 — Row 3
  directions: ExecDirectionRow[];
  governance: ExecGovernanceBlock | null;
  standards: ExecStandardsBlock | null;

  // Pack 5 — Row 2.55 / 2.6 / 2.7
  economic_effect: ExecEconomicEffectBlock | null;
  bp_tracker: ExecBPBlock | null;
  tax_contribution: ExecTaxBlock | null;

  available_years: number[];
  available_sectors: ExecAvailableSector[];
}

const BASE = "/dashboard/executive";

export async function getExecutiveDashboard(
  year: number,
  sectors?: string[],
  bpMetric?: string,
): Promise<ExecutiveDashboardData> {
  const params: Record<string, any> = {};
  if (sectors && sectors.length) {
    params.sectors = sectors;
  }
  if (bpMetric) {
    params.bp_metric = bpMetric;
  }
  const { data } = await apiClient.get<ExecutiveDashboardData>(`${BASE}/${year}`, {
    params,
    paramsSerializer: { indexes: null },
  });
  return data;
}


// ═══ Pack 7.36 — Directions drill modal ═══

export interface ExecDirectionDrillProject {
  id: string;
  title: string;
  status: string;       // init/new/active/review/done
  due_date: string | null;
  progress_percent: number;
  is_overdue: boolean;
  assignee_name: string | null;
}

export interface ExecDirectionDrillTask {
  id: string;
  title: string;
  status: string;
  due_date: string | null;
  progress_percent: number;
  is_overdue: boolean;
  assignee_name: string | null;
  priority: string;     // low/medium/high
}

export interface ExecDirectionDrillCompany {
  company_id: string;
  company_name: string;
  sector: string;       // mining/oilgas/energy/transport/other
  projects_total: number;
  projects_done: number;
  tasks_total: number;
  tasks_done: number;
  tasks_overdue: number;
  projects: ExecDirectionDrillProject[];
  tasks: ExecDirectionDrillTask[];
}

export interface ExecDirectionDrillResponse {
  direction_id: string;       // code, e.g. "esg"
  direction_label: string;
  direction_color: string;
  progress_pct: number;
  companies_count: number;
  projects_total: number;
  projects_done: number;
  tasks_total: number;
  tasks_done: number;
  tasks_overdue: number;
  assignees_count: number;
  companies: ExecDirectionDrillCompany[];
}

/**
 * Fetch detailed breakdown for one direction (Pack 7.36 drill modal).
 *
 * @param year — optional portfolio year filter; null/undefined fetches all
 */
export async function fetchDirectionDrill(
  directionCode: string,
  year?: number,
): Promise<ExecDirectionDrillResponse> {
  const params: Record<string, any> = {};
  if (year != null) params.year = year;
  const { data } = await apiClient.get<ExecDirectionDrillResponse>(
    `${BASE}/directions/${directionCode}`,
    { params },
  );
  return data;
}

File: backend/app/schemas/executive_dashboard.py
"""
backend/app/schemas/executive_dashboard.py — Pydantic схемы для Executive Dashboard.

Pack 1: Row 0 (topbar) + Row 1 (Исполнение задач) + bottom metrics.
Pack 2: + Row 2 — ratings (4 ring cards + table) + execution chart (bar chart).
Pack 3: + Row 2.5 — Финансы · МСФО.
Pack 4: + Row 3 — Направления · Корпуправление · Стандарты.
Pack 5: + Row 2.55 — Экономический эффект.
        + Row 2.6  — BP-трекер портфеля.
        + Row 2.7  — Налоговый вклад в бюджет.
"""
from __future__ import annotations
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ─────────────────────────── Pack 1 ──────────────────────

class ExecCompanyInSector(BaseModel):
    company_id: UUID
    name: str
    pct: int
    board_id: Optional[UUID] = None
    task_total: int = 0
    task_done: int = 0


class ExecSectorRow(BaseModel):
    id: str
    label: str
    color: str
    companies_total: int
    companies_active: int
    avg_pct: int
    companies: List[ExecCompanyInSector]


class ExecBottomMetrics(BaseModel):
    proj_count: int
    task_count: int
    done_proj: int
    done_tasks: int
    deferred_proj: int
    deferred_tasks: int
    avg_completion: int


class ExecAvailableSector(BaseModel):
    id: str
    label: str
    color: str


# ─────────────────────────── Pack 2 — Ratings ──────────────────────

class ExecRingCard(BaseModel):
    label: str
    rated_count: int
    total: int
    not_covered: int
    accent: str
    score: Optional[int] = None
    delta_2024: int = 0


class ExecRatingCell(BaseModel):
    rating: Optional[str] = None
    outlook: Optional[str] = None
    score: Optional[str] = None
    rated_at: Optional[str] = None
    report_url: Optional[str] = None


class ExecRatingRow(BaseModel):
    company_id: UUID
    name: str
    fitch: Optional[ExecRatingCell] = None
    sp: Optional[ExecRatingCell] = None
    moodys: Optional[ExecRatingCell] = None
    sf: Optional[ExecRatingCell] = None
    sp_esg: Optional[ExecRatingCell] = None
    cdp: Optional[ExecRatingCell] = None


class ExecRatingsBlock(BaseModel):
    ring_cards: List[ExecRingCard]
    rows: List[ExecRatingRow]
    rated_total_unique: int
    overall_total: int


# ─────────────────────────── Pack 2 — Execution chart ──────────────────────

class ExecExecutionRow(BaseModel):
    company_id: UUID
    name: str
    pct: int           # факт: % завершённых задач
    plan_pct: int = 0  # план: % задач, чей дедлайн уже наступил (≤ сегодня)
    sector: str


# ═════════════════════════ Pack 4 — Row 3 ═════════════════════════

class ExecDirectionRow(BaseModel):
    id: str
    label: str
    color: str
    projects_total: int
    projects_done: int
    tasks_total: int
    tasks_done: int
    progress_pct: int


# Pack 7.36 — drill modal schemas for "По направлениям" block

class ExecDirectionDrillProject(BaseModel):
    """Один проект в drill-модалке направления."""
    id: UUID
    title: str
    status: str  # init/new/active/review/done
    due_date: Optional[str] = None  # ISO date "YYYY-MM-DD"
    progress_percent: int = 0
    is_overdue: bool = False
    assignee_name: Optional[str] = None


class ExecDirectionDrillTask(BaseModel):
    """Одна задача в drill-модалке направления."""
    id: UUID
    title: str
    status: str
    due_date: Optional[str] = None
    progress_percent: int = 0
    is_overdue: bool = False
    assignee_name: Optional[str] = None
    priority: str = "medium"


class ExecDirectionDrillCompany(BaseModel):
    """Аггрегация по одной компании в drill-модалке направления."""
    company_id: UUID
    company_name: str
    sector: str  # mining / oilgas / energy / transport / other
    projects_total: int
    projects_done: int
    tasks_total: int
    tasks_done: int
    tasks_overdue: int
    projects: List[ExecDirectionDrillProject]
    tasks: List[ExecDirectionDrillTask]


class ExecDirectionDrillResponse(BaseModel):
    """Ответ для GET /dashboard/executive/directions/{code}."""
    direction_id: str   # code, e.g. "esg"
    direction_label: str
    direction_color: str
    progress_pct: int   # tasks-based, same as ExecDirectionRow
    companies_count: int
    projects_total: int
    projects_done: int
    tasks_total: int
    tasks_done: int
    tasks_overdue: int
    assignees_count: int
    companies: List[ExecDirectionDrillCompany]


class ExecGovernanceCompany(BaseModel):
    company_id: UUID
    name: str
    sector: str
    score: int
    score_pct: int
    board_size: int
    independent_count: int
    women_count: int
    indep_pct: int
    women_pct: int


class ExecGovernanceBlock(BaseModel):
    total_companies: int
    avg_score: int
    top_score: int
    avg_indep_pct: int
    avg_women_pct: int
    top_companies: List[ExecGovernanceCompany]


class ExecStandardsRing(BaseModel):
    done: int
    active: int
    init: int
    not_started: int
    pct: int


class ExecStandardsAttention(BaseModel):
    company_id: UUID
    name: str
    sector: str
    ifrs_status: str
    forensic_status: str
    gaps: List[str]


class ExecStandardsBlock(BaseModel):
    total_companies: int
    ifrs: ExecStandardsRing
    forensic: ExecStandardsRing
    attention_list: List[ExecStandardsAttention]


# ═════════════════════════ Pack 5 — Row 2.55 / 2.6 / 2.7 ═════════════════════════

# ─── Block 1 (Pack 5): Экономический эффект ───
class ExecEEKpi(BaseModel):
    """4 KPI band — Экономический эффект."""
    realized_sum: float       # факт (UZS)
    planned_sum: float        # план (UZS)
    pipeline_sum: float       # план - факт
    conversion_pct: int       # realized / planned * 100
    done_count: int
    active_count: int
    total_count: int
    has_data: bool            # False = empty state


class ExecEEProject(BaseModel):
    """Один проект с экономическим эффектом."""
    project_id: UUID
    title: str
    company_name: str
    sector: str
    direction: Optional[str] = None
    status: str
    planned_value: float
    realized_value: float
    pct_realized: int
    unit: str = "млрд сум"


class ExecEconomicEffectBlock(BaseModel):
    """Полный блок Эконом. эффекта."""
    year: int
    kpi: ExecEEKpi
    top_projects: List[ExecEEProject]


# ─── Block 2 (Pack 5 → Pack 7.27): BP-трекер (Performance Spine 1:1 from legacy) ───
class ExecBPCompanyRow(BaseModel):
    """One company in the Performance Spine."""
    company_id: UUID
    name: str
    sector: str
    plan_value: float                    # plan (or prev_year value in YoY mode)
    fact_value: float                    # fact of selected year
    # Pct of plan/prev — None when ratio is meaningless (signed metric edge cases)
    pct: Optional[float] = None          # As fraction (1.0 = baseline)
    display_pct: Optional[int] = None    # Rounded integer % for display, None when not meaningful
    display_label: Optional[str] = None  # Short label (e.g. '↑ восст.', '↓ убыток', '×3.2')
    display_label_full: Optional[str] = None  # Full label (tooltip)
    delta: Optional[float] = None        # cur - ref (raw, signed)
    cls: str = "warn"                    # 'ok' | 'warn' | 'bad'
    note: Optional[str] = None           # 'recovery' | 'loss' | None


class ExecBPBlock(BaseModel):
    """Full BP-tracker block payload — mirrors legacy _execBPData shape."""
    year: int
    prev_year: int
    metric: str                          # 'revenue' | 'ebitda' | 'profit'
    metric_label: str
    standard: str                        # 'BP' | 'NSBU' | 'IFRS' (source of fact)
    mode: str                            # 'plan-fact' | 'yoy' | 'empty'
    head_sub: str                        # subtitle text
    is_signed_metric: bool = False       # True for profit/ebitda

    # Like-for-like totals used for headline percentage
    plan_total: float = 0                # legacy: sumPlan (raw sum, all rows)
    fact_total: float = 0                # legacy: sumFact

    sum_plan_ll: float = 0               # like-for-like sums (only pairs)
    sum_fact_plan_ll: float = 0
    sum_prev_ll: float = 0
    sum_fact_ll: float = 0

    # Headline number
    overall_pct: Optional[float] = None  # ratio (1.0 = baseline) — None when not meaningful
    prev_overall_pct: Optional[float] = None  # ratio from year-2 vs year-1
    overall_delta: Optional[float] = None     # for signed metrics — absolute delta
    overall_label: Optional[str] = None       # for signed metrics — text label

    # Performance Spine: ALL classified companies (sorted by pct desc by backend)
    rows: List[ExecBPCompanyRow]

    # Counters for footer distribution bar
    on_target: int = 0
    attention: int = 0
    behind: int = 0
    total_count: int = 0
    with_pct_count: int = 0

    available_metrics: List[str] = ["revenue", "ebitda", "profit"]

    # Year-fallback: если за запрошенный год данных нет, показываем последний
    # год с данными (year выше), а requested_year хранит исходно выбранный.
    requested_year: Optional[int] = None


# ─── Block 3 (Pack 5): Налоговый вклад ───
class ExecTaxKpi(BaseModel):
    """Сумма налогов (UZS, текущая шкала)."""
    income_tax: float          # Налог на прибыль (sum, в млрд сум)
    vat: float                 # НДС (revenue × 12%, в млрд сум)
    total: float               # income_tax + vat (в млрд сум)
    yoy_total_pct: Optional[float] = None
    yoy_income_tax_pct: Optional[float] = None
    yoy_vat_pct: Optional[float] = None
    budget_share_pct: Optional[float] = None  # % бюджета РУ
    budget: Optional[float] = None             # годовой бюджет РУ для сравнения (млрд)
    vat_is_estimate: bool = True               # Pack 7.9h: НДС оценочный (revenue×12%)


class ExecTaxTopPayer(BaseModel):
    company_id: UUID
    name: str
    sector: str
    amount: float              # сумма налога+НДС от компании (в млрд сум)
    share_pct: int             # доля от total


class ExecTaxBlock(BaseModel):
    year: int
    prev_year: int
    has_data: bool
    standard: str
    cos_count: int
    missing_companies: List[str] = []  # Pack 7.9h: компании без NSBU PL за год
    kpi: ExecTaxKpi
    top_payers: List[ExecTaxTopPayer]
    requested_year: Optional[int] = None  # year-fallback: исходно выбранный год


# ─────────────────────────── Top-level payload ──────────────────────

class ExecutiveDashboardData(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    year: int
    total_companies: int
    title_main: str = "Программа трансформации государственных предприятий"
    title_sub: str

    # Row 1
    row1_title: str = "Исполнение задач Ожиданий Акционера"
    row1_subtitle: str
    sectors: List[ExecSectorRow]
    bottom_metrics: ExecBottomMetrics

    # Row 2 (Pack 2)
    ratings: Optional[ExecRatingsBlock] = None
    execution_chart: List[ExecExecutionRow] = []
    avg_execution_pct: int = 0

    # Row 3 (Pack 4)
    directions: List[ExecDirectionRow] = []
    governance: Optional[ExecGovernanceBlock] = None
    standards: Optional[ExecStandardsBlock] = None

    # Pack 5: Row 2.55 / 2.6 / 2.7
    economic_effect: Optional[ExecEconomicEffectBlock] = None
    bp_tracker: Optional[ExecBPBlock] = None
    tax_contribution: Optional[ExecTaxBlock] = None

    # Filters state
    available_years: List[int]
    available_sectors: List[ExecAvailableSector]
